controller/controller.go

```diff
@@ -113,65 +82,10 @@ type controller struct {
 	Config
 }
 
-// Unit represents a systemd unit file.
-type Unit struct {
-	// Name is something like "appd@.service". It needs to be extended using the
-	// slice ID before submitting to fleet.
-	Name string
-
-	// Content represents normal systemd unit file content.
-	Content string
-}
-
-// Request represents a controller request. This is used to process some action
-// on the controller.
-type Request struct {
-	// Group represents the plain group name without any slice expression.
-	Group string
-
-	// SliceIDs contains the IDs to create. IDs can be "1", "first", "whatever",
-	// "5", etc..
-	SliceIDs []string
-
-	// Units represents a list of unit files that is supposed to be extended
-	// using the provided slice IDs.
-	Units []Unit
-}
-
-var unitExp = regexp.MustCompile("@.service")
-
-// ExtendSlices extends unit files with respect to the given slice IDs. Having
-// slice IDs "1" and "2" and having unit files "foo@.service" and
-// "bar@.service" results in the following extended unit files.
-//
-// 	 foo@1.service
-// 	 bar@1.service
-// 	 foo@2.service
-// 	 bar@2.service
-//
-func (r Request) ExtendSlices() (Request, error) {
-	if len(r.SliceIDs) == 0 {
-		return r, nil
-	}
-	newRequest := Request{
-		SliceIDs: r.SliceIDs,
-		Units:    []Unit{},
-	}
-
-	for _, sliceID := range r.SliceIDs {
-		for _, unit := range r.Units {
-			newUnit := unit
-			newUnit.Name = unitExp.ReplaceAllString(newUnit.Name, fmt.Sprintf("@%s.service", sliceID))
-			newRequest.Units = append(newRequest.Units, newUnit)
-		}
-	}
-
-	return newRequest, nil
-}
-
-func (c controller) Submit(req Request) (*task.Task, error) {
-	if len(req.Units) == 0 {
-		return nil, maskAnyf(invalidArgumentError, "units must not be empty")
+func (c controller) Submit(req api.Request) (*task.Task, error) {
+	ok, err := validator.ValidateRequest(req)
```

Comment:
Why is the validator not simply part of the controller? Then we can prevent all this restructuring. Looks like validation is a controller capability. Just put the source of the validation package into the controller package. What do you think?